### PR TITLE
Update frontend UI design

### PR DESCRIPTION
**Pull request by ericmatte**
Hey @sabeechen ! Well, this took me a lot longer than I thought it would to do this, but here it is!!
Hope you will like it 😄 

### Description of the proposed changes

This PR is a complete revamp of the frontend's design and feel:

image

<details>
<summary><b>Show dark theme example</b></summary>
<br>

image

---

</details>

<details>
<summary><b>Show alternative theme example</b></summary>
<br>

image

---

</details>

This design is a lot closer to the HA design, while maintaining the overall style of this project and MaterializeCSS.

#### Other changes

- Better responsiveness overall
- Frontend now uses the same navbar for ingress and non-ingress views, with some minor tweaks:
  - `ingress`: No hamburger menu + Actions always visible in the top-right
  - `non-ingress`: Hamburger menu on mobile
- Snapshots card revamp:
  - Now show more info
  - Will be disabled while `isPending == true` + a loading animation
  - Are now clickable and will show the details modal directly
  - Snapshot details modal now contains all snapshot actions
- Mobile menu revamp
- Most modals have now a fixed footer
- Update favicon to the new style
- Moved footer to an `About` modal
- Cleaner style for the statistics sidebar
- Extract error messages from `working.jinja2` into their own template (`error-messages.jinja`) for better readability
- Like HA, I am using CSS `var` for easier theming
- Cleanup some code/css & errors

### Screenshots

<details>
<summary><b>Snapshot card</b></summary>
<br>

#### Example (partial, retained + password)
image

#### `isPending` (non-clickable, animation loading)
image

---

</details>

<details>
<summary><b>Snapshot details modal with actions</b></summary>
<br>

image

---

</details>

<details>
<summary><b>Snapshots Statistics sidebar</b></summary>
<br>

image

---

</details>

<details>
<summary><b>New about modal</b></summary>
<br>

image

image

---

</details>

<details>
<summary><b>Updated mobile sidenav</b></summary>
<br>

image

---

</details>

<details>
<summary><b>Error message style</b></summary>
<br>

Instead of a full red background, I've opted for a standard card background, with the title in read:

image

---

</details>

### What to test

- [ ] Go through all the views:
  - [ ] Every feature should work as expected
- [ ] Test all snapshots actions & views
- [ ] Make sure errors messages are displayed properly
- [ ] Play with the background and accent colors:
  - [ ] UI should looks good in any/most cases

Cheers!

### CODE REVIEW

File: hassio-google-drive-backup/backup/static/js/theme.js
Comment (ericmatte) [inline hassio-google-drive-backup/backup/static/js/theme.js:141]: Now, all custom CSS is in `static.css`; and this function only sets the CSS variables.

File: hassio-google-drive-backup/backup/static/js/settings.js
Comment (ericmatte) [inline hassio-google-drive-backup/backup/static/js/settings.js:173]: The view have two `{id}` to replace, and the `string.replace` javascript function only replace one by default; so I call it twice to fix it.

File: .vscode/launch.json
Comment (ericmatte) [inline .vscode/launch.json:44]: That property has been renamed in the update VS-Code python extension.

**Review by sabeechen — CHANGES_REQUESTED** [review]
Man, I have taken FOREVER to get back to you.  I have some good excuses this time, I got coronavirus (not too bad, but don't recommend it), got ANOTHER cold, and had to take care of some family stuff.  Regardless, I'm sorry it took so long.  Please don't take that as some indication that I disapprove of this in any way.  I only have a few small things and maybe one medium size thing to push back on.
In general, Its wonderful to have someone else looking over this code.  I'm not super familiar with all the things CSS/HTML can do so often I just resort to hacking at things until they work, and I noticed all the little cleanup you did along the way with this change.  Thank you for lending some of your wisdom in that way.

The only two big items I think need to be addressed (other than than little things in the comments) are:
*  The footer needs to stay prominent on the addon main pages, as it is the BMC button there thats providing the bulk of the financial support for the project.
* I really like the new snapshot card and its a big improvement over the old one.  I'm less confidant about the mechanics of a 3 column layout.  Many users have long custom snapshot names where the date or otherwise relevant part of the name would get truncated from the current size.  I was thinking to get around this we could:
  * (probably fine) Use a 2-column layout instead for large format screens.  This still has the downside that snapshots aren't in an "obvious" date-order, but I think there would be enough room for most user's custom names.
  * (Better) Use a single column layout, but switch to 2 column if there are more snapshots than would fit in the vertical page space without scrolling.  I'm not sure if there is an elegant way to make materialize's column rules do this.

To support the changes you've made, I'm going to need to also update a laundry list of things.  I don't think any of this should block this submission but they're things I'll need to do before merging this into main and making a release.  Again I'm not asking you to do any of this, its more of a list of "to-do's" for me before making a release of these changes.  I'll dedicate the dev branch to implementing these for the time being, and make bug fixes elsewhere.
* The new snapshot card may need to have an ordinal on it that reflects its date order, or something else that makes it clear which is before/after another.  I'll probably experiment with a few different ways of showing that depending on what the final column layout looks like.
* The server UI page some rework to have it correctly parse the user's accent/bg color when they get redirected.  I might need to do some conditional styling based on the version of the addon being redirect to make the UI match up correctly.  Maybe.
* Snapshot names that get truncated will need to have the full name displayed in a tooltip (or somewhere else) if they get truncted in the new snapshot card.
* Some choices of background/accent color can result in UI elements becoming invisible.  Its kind of a pedantic thing to support, but I'd like to shift the color choices around to at least make the UI usable even if you pick ridiculous colors.
* For some reason textarea elements aren't reflecting the chosen foreground color.  I think this is a pre-existing problem.
* Drop shadows aren't very visible using a dark theme, which makes modals hard to use and crads difficult to see.  I think this too is a pre-existing problem.
* Now that the detail dialog is much more prominent, I need to do some work to make it more sexy.  Its been an afterthought until now that I'm sure most users aren't even aware of.  I'll probably also make the action button do soemthing inline within the dialog instead of opening a modal on top of the modal.
* The "?" info icon on the snapshot dialog can be removed now that the whole thing opens the detail dialog.
* The snapshot's icon should correspond to the snapshot's state (eg Backed Up, in HA, etc).  I think that woudl make sense now that the icon is something different if the snapshot is pending.
* The snapshot detail modal needs to do something smart if the snapshot's state changes in a significant way while the modal is open (previously the action menu could just hide items it needed to in real time when that happens).
* The source icon (eg "sd_card" or "cloud") need to be defined in code by snapshtot source in order to support some future improvements.
* AioHttp request logs have started showing up in the debug logs (and maybe the addon logs too?), configure the logger to ignore them.
* Test each error card individually with whatever theme changes end up happening, to make sure no UI elements get hidden or look off.
* Now that buttons are less visually distinct form the rest of the UI (eg no colored background) I think each one should have an icon to help them stand out a little.  This is something I think the HA UI gets wrong.

File: hassio-google-drive-backup/backup/static/js/theme.js
Comment (ericmatte) [reply hassio-google-drive-backup/backup/static/js/theme.js:141]: Yeah! And it is actually what Home Assistant also uses ;)

File: hassio-google-drive-backup/backup/static/js/scripts.js
Comment (ericmatte) [reply hassio-google-drive-backup/backup/static/js/scripts.js:92]:
```suggestion
```

File: hassio-google-drive-backup/backup/static/layouts/partials/error-messages.jinja2
Comment (ericmatte) [reply hassio-google-drive-backup/backup/static/layouts/partials/error-messages.jinja2:1]: > I'll need to do a little more testing of these before I'm confidant everything works correctly, but at a first pass everything looks good for submission.

Great idea! While I double checked everything, it is a good idea to do more testing on that.
I've found out that some errors were using the wrong id or duplicated ids, so I fixed and tested those.
Also in the errorMessage macro, I put the error id in both the `id` and `class` fields, since some of the code was calling them using their id and other part using their name in the class field:
```html
<div class="card default-hidden error_card {{ id }}" id="{{ id }}">
```

File: hassio-google-drive-backup/backup/static/layouts/partials/modals/settings.jinja2
Comment (ericmatte) [reply hassio-google-drive-backup/backup/static/layouts/partials/modals/settings.jinja2:517]: Yeah; display flex is awesome and I use it all the time for responsive UI.

File: hassio-google-drive-backup/backup/static/layouts/partials/modals/snapshot/new.jinja2
Comment (ericmatte) [reply hassio-google-drive-backup/backup/static/layouts/partials/modals/snapshot/new.jinja2:54]: Thanks! I tried to make it simpler to use and to better show the information.

File: hassio-google-drive-backup/backup/static/layouts/partials/modals/snapshot/new.jinja2
Comment (ericmatte) [inline hassio-google-drive-backup/backup/static/layouts/partials/modals/snapshot/new.jinja2:1]: > To support the changes you've made, I'm going to need to also update a laundry list of things. I don't think any of this should block this submission but they're things I'll need to do before merging this into main and making a release. Again I'm not asking you to do any of this, its more of a list of "to-do's" for me before making a release of these changes. I'll dedicate the dev branch to implementing these for the time being, and make bug fixes elsewhere.

I could probably do some of it!
Like:
- Snapshot names that get truncated will need to have the full name displayed in a tooltip (or somewhere else) if they get truncted in the new snapshot card.
- The "?" info icon on the snapshot dialog can be removed now that the whole thing opens the detail dialog.
- Drop shadows aren't very visible using a dark theme, which makes modals hard to use and cards difficult to see. I think this too is a pre-existing problem.

About the drop shadows, they works ok when you use a background color slightly paler than black. But on a pitch black background, you are right; they are just not visible. I didn't want to use a white drop shadow on darker backgrounds, as it made the UI less appealing. But I could probably make something in-between for that case.

**Review by ericmatte — COMMENTED** [review]
Hey @sabeechen,

Don't worry, I was in no rush to merge this PR. Glad to hear back from you, and to know that you are now doing better after catching covid and everything.

File: hassio-google-drive-backup/backup/static/layouts/base-server.jinja2
Comment (sabeechen) [reply hassio-google-drive-backup/backup/static/layouts/base-server.jinja2:36]: - Having the smaller, one-line footer I think is perfect.
  - I think it should read <code>© 2021 Stephen Beechen - If you like this project, keep it going and <kbd>Buy me a Coffee<kbd></code>
  - I could go either way with it being sticky.  So your call, I think you have a better eye for such things :)
- I agree having the other info in the about dialog is much better, I like the dialog and I'll add more to it in the future.
- Double yes to having a contributor card in the about dialog.  If you don't do it, I will.

File: hassio-google-drive-backup/backup/static/layouts/partials/modals/snapshot/new.jinja2
Comment (sabeechen) [reply hassio-google-drive-backup/backup/static/layouts/partials/modals/snapshot/new.jinja2:1]: Oh yes I'd absolutely appreciate the help.  I just don't think any of that work should block this PR (its big plenty big already).

I'm also not sure what exactly to do about the drop shadows, since I agree the pure white ones look *awful* on pure black. I was meaning to fiddle around with the colors to see if I can get something to look nice.

**Comment by sabeechen** [comment]
I'm pretty excited to get this out. I think people are going to like it.  I think this addon already has a pretty slick UI (compared to other addons) and you're setting the bar even higher.
Fair warning though, change aversion is a real thing too, so expect a little hate also.  But they can fuck right off.

**Comment by ericmatte** [comment]
Hey @sabeechen, just to let you know that I will probably work on finishing this PR this weekend, or in the beginning of the next week.

File: hassio-google-drive-backup/backup/static/layouts/base-server.jinja2
Comment (ericmatte) [reply hassio-google-drive-backup/backup/static/layouts/base-server.jinja2:36]: @sabeechen what do you think of this footer?

image

Comment (sabeechen) [reply hassio-google-drive-backup/backup/static/layouts/base-server.jinja2:36]: Looks good to me!

File: hassio-google-drive-backup/backup/static/css/static.css
Comment (ericmatte) [inline hassio-google-drive-backup/backup/static/css/static.css:2]: > * I really like the new snapshot card and its a big improvement over the old one.  I'm less confidant about the mechanics of a 3 column layout.  Many users have long custom snapshot names where the date or otherwise relevant part of the name would get truncated from the current size.  I was thinking to get around this we could:
>   
>   * (probably fine) Use a 2-column layout instead for large format screens.  This still has the downside that snapshots aren't in an "obvious" date-order, but I think there would be enough room for most user's custom names.
>   * (Better) Use a single column layout, but switch to 2 column if there are more snapshots than would fit in the vertical page space without scrolling.  I'm not sure if there is an elegant way to make materialize's column rules do this.

@sabeechen I've added a small tooltip if you hover your mouse over the truncated text.

As for the 3 column layout, I think that we could leave it like it is. I've made the size of each card similar to what HA is doing:

image

vs

image

Also note that I'm using `col s12 m12 l6 xl4`, which means that we only get 3 columns on xl screens.

> Many users have long custom snapshot names where the date or otherwise relevant 

That is why I've added a text which says when the screenshot was taken 😄 : 
image

---

Here is a screenshot if I switch to 2 columns on xl screens:

image

I have a preference for the 3 columns layout, but 2 also looks clean.

So if you still prefer 2, I will switch no problem. Let me know what you think!

File: hassio-google-drive-backup/backup/static/layouts/base-server.jinja2
Comment (ericmatte) [reply hassio-google-drive-backup/backup/static/layouts/base-server.jinja2:36]: @sabeechen here is the updated about modal with contributors:

image

File: hassio-google-drive-backup/backup/static/layouts/partials/modals/snapshot/new.jinja2
Comment (ericmatte) [reply hassio-google-drive-backup/backup/static/layouts/partials/modals/snapshot/new.jinja2:1]: - [x] Snapshot names that get truncated will need to have the full name displayed in a tooltip (or somewhere else) if they get truncted in the new snapshot card.
- [x] The "?" info icon on the snapshot dialog can be removed now that the whole thing opens the detail dialog.
- [x] Drop shadows aren't very visible using a dark theme, which makes modals hard to use and cards difficult to see. I think this too is a pre-existing problem.

---

So for the drop shadows, what I did is that I use white shadows only on pitch-black background. The rest of the time the shadows are black:

### Absolute black:

image

image

### Lighter black:

image

image

### White:

image

image

**Review by ericmatte — COMMENTED** [review]
So @sabeechen I've fixed about everything that I wanted to do with this PR.

Just make sure everything looks good for you, and I think we could merge this.

One thing: I've kept the 3 columns layout, with this [explanation](https://github.com/sabeechen/hassio-google-drive-backup/pull/332#discussion_r589098747). Just let me know what you think (I can change it if you prefer).

**Review by sabeechen — APPROVED** [review]
Thanks for the fixes, I think this looks good and its in a state sufficient to merge it in.  It's kind of difficult for me to see which comments are resolved/unresolved, so I'm just going to address everything here. 

* **Snapshot Column Layout** I'm probably going to switch it to a 2 column layout, but I'd like to play around with a few different options before deciding on that.  In general I'd like to be consistent with the HA UI, but in many cases I think they've leaned to heavy on style vs substance.  More than just the date is important, since some people put a *ton* of details in the snapshot name 
* **Contributor UI**  That looks great!  I'm going to make some additions to it, but they'll be a surprise :)  Stay tuned!
* **Black background drop shadow**  It looks much better with the white drop shadow.  You're right that it only seems to be a problem on "very close to actual black" backgrounds.  I'm going to be playing around with a bunch of the generated CSS rules and I'll see if I can improve on it any (some fixes form this PR, but also things I've been meaning to get around to for a long time)

So then here is my plan for this PR:
* I'll merge it into dev now.
* I'll merge some bug fixes and features I've been working on in a separate branch into dev as well.  I was hoping to make a separate release with just those, but it took me longer to complete so I'll just make them all in a big release.
* I'll start working on the list of little changes I needed to make in random places
* v1.104.0 will feature your new UI.  I'd estimate I'll have it out in within 2 weeks as I need to make those changes and I'm going to let it run on my dev machine under a normal useage scenario for a while to find any issues we might have overlooked.

**Comment by ericmatte** [comment]
@sabeechen thank you for the great review; can't wait to see the new release in my own Home Assistant!

Also, the README.md will probably need a new screenshot :)

**Comment by sabeechen** [comment]
Its on the list :)

**Comment by sabeechen** [comment]
I've implemented all of the changes I listed earlier and submitted them to Dev.  I need to let it bake on my test instance for a while and a I have a few more features to implement, but that is otherwise all that needs to be done before its released.  Only really difficult part was redesigning the details dialog which is much, *much* better now.

I've also spent some time cleaning up the code and automating releases.  All the automated tests now work and I've set thigns up so that any commits pushed to dev automatically build and deploy the addon and the authentication server to a staging instance of the addon that talks to [https://dev.habackup.io](https://dev.habackup.io) instead of [https://habackup.io](https://habackup.io).  You can run this staging version of the addon by adding this repo to Home assistant:

[https://github.com/sabeechen/hgdb-dev-staging](https://github.com/sabeechen/hgdb-dev-staging)

This installs an instance of the addon that is a separate "app" as far as Google is concerned so it can't see any snapshots maintained by the "Production" addon.  Any new commits to dev show up as a new version of the addon in Home Assistant.  It can be run side-by-side with the production addon, they'll both try to create snapshots so I wouldn't really recommend it.

If you'd like to run the addon with your changes or would like to look it over to see if I messed up anything you did, that repo is the way.  It woudl be good to have a second pair of eyes on it running "for real".

**Comment by sabeechen** [comment]
Other note: the staging addon right now only builds the amd64 version of the addon since thats what I use.  If you'd like to use it to run on another platform (eg raspberry pi), let me know and I'll have it publish those images as well as part of the automated deployment.

**Comment by ericmatte** [comment]
@sabeechen good idea! A staging environment is great.

I am using a Raspberry Pi 4 64bits for my HA instance.

I did try to install it, but I'm receiving this 404 error:

> 21-03-16 14:01:54 ERROR (SyncWorker_3) [supervisor.docker.interface]
Can't install sabeechen/hassio-google-drive-backup-aarch64_staging:0.103.1.staging.26 -> 404 Client Error for http+docker://localhost/v1.40/images/create?tag=0.103.1.staging.26&fromImage=sabeechen%2Fhassio-google-drive-backup-aarch64_staging: Not Found ("pull access denied for sabeechen/hassio-google-drive-backup-aarch64_staging, repository does not exist or may require 'docker login': denied: requested access to the resource is denied").

**Comment by sabeechen** [comment]
Make sure home assistant sees v0.103.1.staging.27 or higher for the addon (you might need to refresh the addon-store) then give it another shot.  I just had the builder upload all platform images including aarch64 (64 bit arm for the rpi).

**Comment by ericmatte** [comment]
@sabeechen I indeed needed to refresh the addon-store. It is working now! I will do a quick check/tests on it et let you know if there is anything weird.

**Comment by sabeechen** [comment]
Took me forever to get the other features associated with this release done, but I've finally released the new UI in v0.104.1 (latest is v0.104.2).  People seem to love it so far.  Thanks for the huge contribution!

**Comment by ericmatte** [comment]
Great news! @sabeechen it was my pleasure :smi

**Pull request merged**
